dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/alias/InclusionBasedPointerAnalysis.java

```diff
@@ -0,0 +1,816 @@
+        // This could enable more communications, but replacement.
+        for (final Offset<Variable> load : old.loads) {
+            addInto(replacement.base.loads, join(load, replacement.offset, replacement.alignment));
+            load.base.seeAlso.add(replacement.base);
+        }
+        for (final Offset<Offset<Variable>> store : old.stores) {
+            addInto(replacement.base.stores, join(store, replacement.offset, replacement.alignment));
+            store.base.base.seeAlso.add(replacement.base);
+        }
+        old.seeAlso.clear();
+        old.includes.clear();
+        old.loads.clear();
+        old.stores.clear();
+    }
+
+    private void processCommunication(Collection<Offset<Offset<Variable>>> stores, Collection<Offset<Variable>> loads) {
+        logger.trace("{} vs {}", stores, loads);
+        if (loads.isEmpty()) {
+            return;
+        }
+        for (final Offset<Offset<Variable>> store : stores) {
+            for (final Offset<Variable> load : loads) {
+                if (overlaps(load.offset - store.offset, store.alignment, load.alignment)) {
+                    addEdge(store.base.base, load.base, store.base.offset, store.base.alignment);
+                }
+            }
+        }
+    }
+
+    private record Offset<Base>(Base base, int offset, List<Integer> alignment) {
+        @Override public String toString() { return String.format("%s+%d+%sx", base, offset, alignment); }
+    }
+
+    // checks if leftAlignment contains all linear combinations of offset + rightAlignment
+    private boolean includes(int offset, List<Integer> leftAlignment, List<Integer> rightAlignment) {
+        if (totalAlignmentSizes != null) {
+            totalAlignmentSizes.computeIfAbsent(leftAlignment.size() + rightAlignment.size(), k -> new int[1])[0]++;
+        }
+        if (leftAlignment.isEmpty()) {
+            return rightAlignment.isEmpty() && offset == 0;
+        }
+        for (final Integer a : leftAlignment) {
+            if (a < 0) {
+                final int l = absGcd(leftAlignment);
+                final int r = absGcd(rightAlignment);
+                return offset % l == 0 && r % l == 0;
+            }
+        }
+        for (final Integer a : rightAlignment) {
+            if (a < 0) {
+                final int l = absGcd(leftAlignment);
+                final int r = absGcd(rightAlignment);
+                return offset % l == 0 && r % l == 0;
+            }
+        }
+        // FIXME assumes that dynamic indexes used here only have non-negative values.
+        // This cannot be used when a negative alignment occurs, because the analysis would not terminate.
+        if (leftAlignment.size() == 1) {
+            final int alignment = Math.abs(leftAlignment.get(0));
+            for (final Integer a : rightAlignment) {
+                if (a % alignment != 0) {
+                    return false;
+                }
+            }
+            return offset % alignment == 0;
+        }
```

Comment:
The issue talks about dynamic indices being non-negative (static indices are fine and occur from `container_of` constructs),
and for those we cannot check if they are non-negative (we don't know!).

Reply:
The guards before this line test if there is a negative value.  So this line is only reachable if all values are non-negative.  I had to add negative values to pass some tests.  The FIXME is there because I would like to have an algorithm that also works for negative values.
I am trying to fix this issue without loosing the precision we are used to by now.  Including the existing analyses, `IBPA` assumes non-negative dynamic indexes / offsets if reading it from a register, regardless of its original C type, which would oftentimes be signed.
There is an interesting theory behind it, but I don't think I will get this implemented soon.

Reply:
I think as long as the known edge-cases are documented somewhere (e.g., in the form of FIXMEs) I think the current behavior is fine.
There are papers out their that literally say "sound and precise alias analysis of C code is impossible" :).